lib/streamlit/elements/lib/mutable_status.py

```diff
@@ -0,0 +1,164 @@
+
+from __future__ import annotations
+
+import time
+from types import TracebackType
+from typing import Any, List, Optional, Type, cast
+
+from typing_extensions import Literal, TypeAlias
+
+from streamlit.cursor import Cursor
+from streamlit.delta_generator import DeltaGenerator, _enqueue_message
+from streamlit.proto.Block_pb2 import Block as BlockProto
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+States: TypeAlias = Literal["running", "complete", "error"]
+
+
+class MutableStatus(DeltaGenerator):
+    # @contextmanager
+    @staticmethod
+    def _create(
+        parent: DeltaGenerator,
+        label: str,
+        expanded: bool | Literal["auto"] = False,
+        state: States = "running",
+    ) -> MutableStatus:
+        # -> Generator[MutableStatus, None, MutableStatus]:
+        expandable_proto = BlockProto.Expandable()
+        expandable_proto.expanded = True if expanded == "auto" else expanded
+        expandable_proto.label = label or ""
+
+        if state == "running":
+            expandable_proto.icon = "spinner"
+        elif state == "complete":
+            expandable_proto.icon = "check"
+        elif state == "error":
+            expandable_proto.icon = "error"
+
+        block_proto = BlockProto()
+        block_proto.allow_empty = True
+        block_proto.expandable.CopyFrom(expandable_proto)
+
+        # TODO(lukasmasuch): Check if this is the correct delta path even when it's nested.
+        delta_path: List[int] = (
+            parent._active_dg._cursor.delta_path if parent._active_dg._cursor else []
+        )
+
+        status_container = cast(
+            MutableStatus,
+            parent._block(block_proto=block_proto, dg_type=MutableStatus),
+        )
+
+        status_container._delta_path = delta_path
+        status_container._last_proto = block_proto
+        status_container._last_state = state
+        if expanded == "auto":
+            status_container._auto_collapse = True
+
+        # TODO: We cannot use this since it won't support the usage without context manager:
+        # with status_container:
+        #     try:
+        #         yield status_container
+        #     except Exception as ex:
+        #         status_container.update(state="error")
+        #         raise ex
+        return status_container
+
+    def __init__(
+        self,
+        root_container: int | None,
+        cursor: Cursor | None,
+        parent: DeltaGenerator | None,
+        block_type: str | None,
+    ):
+        super().__init__(root_container, cursor, parent, block_type)
+
+        # Initialized in `_create()`:
+        self._last_proto: BlockProto | None = None
+        self._last_state: States | None = None
+        self._delta_path: List[int] | None = None
+        self._auto_collapse: bool = False
+
+    def update(
+        self,
+        *,
+        label: str | None = None,
+        state: States | None = None,
+        expanded: bool | None = None,
```

Comment:
tiny nit: should order of these be switched to match st.status signature?

Reply:
or maybe status is the more important kwarg here so this makes sense 🤔 if we may expanded a kwarg for st.status, it should probably be the same order. I'm OK either way but being consistent seems good